src/integer_value.cpp

```diff
@@ -116,27 +117,26 @@ ValuePtr IntegerValue::mul(Env *env, ValuePtr arg) {
     }
     arg.assert_type(env, Value::Type::Integer, "Integer");
 
+    if (is_zero() || arg->as_integer()->is_zero())
+        return new IntegerValue { 0 };
+
     if (arg.is_bignum()) {
         auto other = arg->as_integer();
         auto result = to_bignum() * other->to_bignum();
         return new BignumValue { result };
     }
 
-    nat_int_t result = to_nat_int_t() * arg.to_nat_int_t();
-    bool overflowed = false;
-    bool same_sign = (to_nat_int_t() ^ arg.to_nat_int_t()) >= 0;
-    if (!same_sign && result > 0)
-        overflowed = true;
-    if (same_sign && result < 0)
-        overflowed = true;
+    auto ll_this = to_nat_int_t();
+    auto ll_arg = arg.to_nat_int_t();
 
-    if (overflowed) {
-        auto other = arg->as_integer();
-        auto result = to_bignum() * other->to_bignum();
-        return new BignumValue { result };
+    auto min_fraction = (NAT_MIN_FIXNUM - (NAT_MIN_FIXNUM % ll_arg)) / ll_arg;
+    auto max_fraction = (NAT_MAX_FIXNUM - (NAT_MAX_FIXNUM % ll_arg)) / ll_arg;
+    if (
+        (ll_this > 0 && ll_arg > 0 && max_fraction <= ll_this) || (ll_this > 0 && ll_arg < 0 && min_fraction <= ll_this) || (ll_this < 0 && ll_arg > 0 && min_fraction >= ll_this) || (ll_this < 0 && ll_arg < 0 && max_fraction >= ll_this)) {
+        return (new BignumValue { to_bignum() })->mul(env, arg);
```

Comment:
I think we should unify this with the other bignum case above. We should either call the mul method on BignumValue or do the multiplication in here. What do you think?

Also the if statement above looks a bit weird formatting-wise or is this just me?

Reply:
I am not sure I would like to unify it, the case above is a faster path and doesn't need us to calculate the fractions, plus I think it is nicely separated.

Reply:
What I meant by this is that in this branch we use
```cpp
(new BignumValue { to_bignum() })->mul(env, arg)
```

In the other faster path above we use 

```cpp
        auto result = to_bignum() * other->to_bignum();
        return new BignumValue { result };
```

I'm not quite sure which variant I prefer. The first one might be more consistent (for example if BignumValue multiplication does more that just multiply (any checks or whatever)). The second variant however is probably a bit faster because we save some checks (like checking for floats and so on). What do you think?

Reply:
I tend to like the consistency of the second one, but I do see your point! In fact, now I am also quite conflicted, @seven1m  any preference? We should align those anyway, the question is which variant we will end up choosing :)

Reply:
Yeah I don't care, I'll let you guys decide. But in the first case, you shouldn't need the `new` keyword since the intermediate value is thrown away:

```c++
return (BignumValue { to_bignum() })->mul(env, arg);
```

Reply:
To clarify, I don't think I have enough info to know which is the better course. I think we just need to pick a path and figure out later if it was a mistake. 😄